### DISCUSSION TITLE
Setting additional IRQ flags for SX126x/SX128x

### DISCUSSION THREAD

**Discussion by GUVWAF** [jgromes/RadioLib#682] [Ideas]
Lately I’ve been looking into the lower layers of Meshtastic, whose firmware roughly works like this. Before transmitting a packet, the device first tries to determine whether it is actively receiving. If not, it performs channel activity detection. Only when the channel is free, it starts transmitting (CSMA/CA).
The former condition works well for SX127x modules by checking the modem status whether a valid header was received. For SX126x and SX128x modules, the same check is implemented, but then using `getIrqStatus()`. I found out that since the `IRQ_HEADER_VALID` flag is not mapped to the IRQ in `startReceiveCommon()`, this actually doesn’t work. Although the CAD procedure afterwards also detects the data part of the packet, it fails to receive it because the module went into standby mode while receiving. This has been mentioned before in #67.

I tested that simply adding `RADIOLIB_SX126X_IRQ_HEADER_VALID` to the IRQ mask works to check whether we are actively receiving and since it will then not perform CAD, the packet can be received successfully.
Do you think it would be good to expose an option for setting additional IRQ flags to the receive methods? Or maybe even a configuration, similar to `setDIOMapping()`?. I’m willing to create a PR for this, but would first like to know your opinion on this. Maybe this is not the right way to do it, which I would then also like to hear.

**Comment by jgromes** [comment]
I'm wondering if it wouldn't be easier to just add the `IRQ_HEADER_VALID` flag to `startReceiveCommon()`. If it doesn't interfere with existing high-level functionality (namely blocking and interrupt receive), then there's proabably no need to manually add the flag via some configuration method.

**Reply by GUVWAF** [reply]
Thanks for the quick reply.
I'm afraid it will interfere with the blocking receive functionality, because the IRQ will fire when only the header has been received, but then it cannot yet read the data. It would be possible to catch this case in the function `receive()`, but I'm not sure if that's desired.
Since the SX126x and SX128x modules don't have a `setIrqAction()` method, it should not interfere with the interrupt receive functionality.

**Reply by GUVWAF** [reply]
Maybe the best way to handle this is to add a parameter (e.g. `bool setIrqAll`) to `startReceiveCommon()` for SX126x and to `startReceive()` for SX128x. This value is true by default and set to false when calling it from the blocking function `receive()`. Then if it is true, the IRQ mask is set to IRQ_ALL, otherwise to RX_DONE, TIMEOUT, CRC_ERR and HEADER_ERR (what it is now). This way users can get the full result from `getIrqStatus()` when using interrupt receive, and it will not affect blocking receive.

What are your thoughts on this?

**Reply by Velocet** [reply]
> I'm afraid it will interfere with the blocking receive functionality, because the IRQ will fire when only the header has been received, but then it cannot yet read the data. It would be possible to catch this case in the function `receive()`, but I'm not sure if that's desired.

Isn't that the intended way? If i read Semtechs explanations correctly you have two ways of dealing with this problem: one blocking and one non-blocking method. But either handles the fifo fill condition a bit different. There is a topic about this at Semtechs Forum if my memory is right...

**Reply by GUVWAF** [reply]
@Velocet It seems wasteful to me to do an additional SPI call for every packet reception just to check the IRQ status if we can proceed with reading the data already. I can't seem to find a post on Semtech's forum on this.

**Reply by jgromes** [reply]
@GUVWAF I tested blocking reception with IRQ_HEADER_VALID added, and while it seems that it does not interfere with functionality, this can be due to the fact that I'm testing on Arduino Uno, which is comparatively slow compared to something like ESP32.

I think the best (easiest and most generic) approach will be to add an argument to `startReceiveCommon` that will allow user to define their own IRQ flags and mask, e.g. `startReceiveCommon(uint32_t timeout, uint16_t irqFlags, uint16_t irqMask)`. The arguments must default to the current values so that we don't introduce a breaking change and will also have to be propagated from `startReceive` methods. I assume `startReceive()` is the method Meshtastic is calling, correct?

**Reply by GUVWAF** [reply]
@jgromes Thanks for testing. That might indeed be the case.

Meshtastic is actually calling `startReceiveDutyCycleAuto()` for SX126x, so it would also need to be added to that method, and propagated to `startReceiveDutyCycle()`. 
Would it be fair to assume that currently a user would "logical AND" the IRQ status result with the mask they're interested in? Then using `IRQ_ALL` for interrupt receive and the current values for blocking receive would not create a breaking change. That would save some parameters; I've just drafted it up [here](https://github.com/GUVWAF/RadioLib/tree/setIrqAll). 

Otherwise, I can proceed with your approach. However, I believe only `irqMask` is needed, right?

**Reply by jgromes** [reply]
> Meshtastic is actually calling startReceiveDutyCycleAuto() for SX126x, so it would also need to be added to that method, and propagated to startReceiveDutyCycle().

Of course, all of the `startReceiveXYZ` methods need to expose this argument.

> Would it be fair to assume that currently a user would "logical AND" the IRQ status result with the mask they're interested in?

I don't think that assumption can be made - the library has no idea why the user is calling a particaular method. The mask is actually one of the arguments passed to `setDioIrqParams`, it is the mask of IRQs that will cause a change in DIOx.

> However, I believe only irqMask is needed, right?

I don't think that assumption can be made either. I would prefere a more generic mathod taht lets the user set arbitrary IRQ and DIOx mask, separate from each other. Perhaps the user is intentionally enabling more interrupts than are routed to the DIOx, but processing only some of them as interrupts and some by polling?

**Reply by GUVWAF** [reply]
That makes sense, thanks for the clarification. I now realize how hard it is to create a generic library ;)
I will proceed with your approach for a PR then.

**Reply by GUVWAF** [reply]
@jgromes While implementing these changes, I observed that using commit 00998b5, setting the SX1262 module to standby fails with error -707, while commit 2b2dc29 works.

**Reply by jgromes** [reply]
@GUVWAF most likely you are trying to use SX126x without TCXO. For those, the TCXO voltage must be set to 0.

**Reply by GUVWAF** [reply]
@jgromes It has a TCXO (it's a RAK4361). I think the default `tcxoVoltage` was accidentally changed in bd25838 (and fixed again in 00998b5), but it was already set to 1.6 in 2b2dc29.
Can it be related to changes in the SPI functions?

**Reply by jgromes** [reply]
I don't think it has anything to do with the low-levle SPI changes, as I tested them on both TCXO and XTAL module. I suggest enabling debug + verbose and taking a look at what exactly is failing.

**Reply by GUVWAF** [reply]
The SPI call for standby gave me:
> CMD	80	
DATW	0	AA

While normally it returns:
> CMD	80	
DATW	0	D2	

I found that if I comment out the call to `SX126x::findChip()` in `SX126x::begin()`, it runs fine.
Edit: this thus means it could not find the correct chip, but I cannot yet see those logs just after startup. The correct SX1262 module is used, I will double-check the wiring later.

**Reply by jgromes** [reply]
I was not able to reproduce this problem. However, as it seems unrelated to the original point of this discussion, please post this as a new issue, including the full verbose SPI transcript.
